src/DependencyGraph/DependencyGraphHelpers.js

```diff
@@ -8,24 +8,89 @@
  */
 'use strict';
 
+const fs = require('fs');
 const path = require('fast-path');
 
 class DependencyGraphHelpers {
-  constructor({ providesModuleNodeModules, assetExts }) {
-    this._providesModuleNodeModules = providesModuleNodeModules;
+  constructor({ roots, providesModuleNodeModules, assetExts }) {
+    this._roots = roots;
+    this._providesModuleNodeModuleDirs = this._resolveHastePackages(providesModuleNodeModules);
     this._assetExts = assetExts;
   }
 
+  /**
+   * An item in the `providesModuleNodeModule` array is an object containing two keys:
+   *  * "name" - the package name
+   *  * "parent" - if a string, it's the name of the parent package of the module in
+   *  the dep tree. If null, it signifies that we should look for this package at
+   *  the top level of the module tree.
+   *
+   * We try to resolve the specified module...but if we can't, we fallback to looking
+   * in "node_modules/[package name]" (this satisfies use cases where the actual
+   * "react-native" package is one of the root dirs, such as when we run tests
+   * or examples).
+   */
+  _resolveHastePackages(packages) {
+    const packagePathForPackage = ({ name, parent }, rootDir) => {
+      let packagePath;
+      if (parent) {
+        if (!Array.isArray(parent)) {
+          parent = [parent];
+        }
+        parent.push(name);
+        packagePath = path.join(
+          rootDir,
+          'node_modules',
+          parent.join(path.sep + 'node_modules' + path.sep)
+        );
+      } else {
+        packagePath = path.join(rootDir, 'node_modules', name);
+      }
+
+      if (packagePath[packagePath.length - 1] === path.sep) {
+        return packagePath.slice(0, -1);
+      } else {
+        return packagePath;
+      }
+    };
+
+    return packages.reduce((hastePackages, p) => {
+      this._roots.forEach(rootDir => {
+        const packagePath = packagePathForPackage(p, rootDir);
+        try {
+          const stats = fs.statSync(packagePath);
+          if (stats && stats.isDirectory()) {
+            hastePackages.push(packagePath);
+          }
+        } catch (e) {
+          // if we don't find the package, let's just default to node_modules/[package name]
+          hastePackages.push(packagePathForPackage({ name: p.name }, rootDir));
+        }
+      });
+      return hastePackages;
+    }, []);
+  }
+
+  /**
+   * This method has three possible outcomes:
+   *
+   * 1) A file is not in "node_modules" at all (some type of file in the project).
+   * 2) The file is in "node_modules", and it's contained in one of the
+   * "providesModuleNodeModules" packages.
+   * 3) It's in "node_modules" but not in a "providesModuleNodeModule" package,
+   * so it's just a normal node_module.
+   */
   isNodeModulesDir(file) {
-    const index = file.lastIndexOf('/node_modules/');
+    const index = file.indexOf('/node_modules/');
```

Comment:
``file.indexOf(`${path.sep}node_modules${path.sep}`);``? that’s so ugly that you might want to put a constant at the top of the file :-)